### PR TITLE
OpenAI-DotNet 7.2.3

### PR DESCRIPTION
**Pull request by chsword**
- Added support for reading RateLimit information from the Headers

### CODE REVIEW

**Review by StephenHodgson — CHANGES_REQUESTED** [review]
Thanks for the PR!

Please fix formatting and update version and release notes

**Comment by chsword** [comment]
> Thanks for the PR!
> 
> Please fix formatting and update version and release notes

I have formatted the code and added the version number and ReleaseNotes.

File: OpenAI-DotNet/Extensions/ResponseExtensions.cs [hunk 1]
```diff
@@ -24,7 +32,10 @@ internal static class ResponseExtensions
 
         internal static void SetResponseData(this BaseResponse response, HttpResponseHeaders headers)
         {
-            if (headers == null) { return; }
+            if (headers == null)
+            {
+                return;
+            }
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Extensions/ResponseExtensions.cs:38]: Revert

File: OpenAI-DotNet/Completions/CompletionResult.cs [hunk 2]
```diff
@@ -47,9 +47,14 @@ public sealed class CompletionResult : BaseResponse
...
-
+        
+        [JsonIgnore]
+        public RateLimits RateLimits { get; set; }
+        
         public override string ToString() => FirstChoice?.ToString() ?? string.Empty;
 
         public static implicit operator string(CompletionResult response) => response.ToString();
+        
+
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Completions/CompletionResult.cs:58]: Remove whitespace

File: OpenAI-DotNet/Completions/CompletionResult.cs [hunk 3]
```diff
@@ -47,9 +47,14 @@ public sealed class CompletionResult : BaseResponse
 
         [JsonIgnore]
         public Choice FirstChoice => Completions?.FirstOrDefault(choice => choice.Index == 0);
-
+        
+        [JsonIgnore]
+        public RateLimits RateLimits { get; set; }
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Completions/CompletionResult.cs:52]: Internal set only

File: OpenAI-DotNet/Chat/ChatResponse.cs [hunk 4]
```diff
@@ -68,6 +68,9 @@ public IReadOnlyList<Choice> Choices
 
         [JsonIgnore]
         public Choice FirstChoice => Choices?.FirstOrDefault(choice => choice.Index == 0);
+        
+        [JsonIgnore]
+        public RateLimits RateLimits { get; set; }
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Chat/ChatResponse.cs:73]: Internal set only

File: OpenAI-DotNet/RateLimits.cs [hunk 5]
```diff
@@ -0,0 +1,37 @@
+﻿namespace OpenAI
+{
+
```
Comment (StephenHodgson) [inline OpenAI-DotNet/RateLimits.cs:3]: Remove whitespace

File: OpenAI-DotNet/RateLimits.cs [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    public class RateLimits
+    {
+
```
Comment (StephenHodgson) [inline OpenAI-DotNet/RateLimits.cs:6]: Remove whitespace

File: OpenAI-DotNet/RateLimits.cs [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        /// <summary>
+        /// The maximum number of requests that are permitted before exhausting the rate limit.
+        /// </summary>
+        public int LimitRequests { get; set; }
```
Comment (StephenHodgson) [inline OpenAI-DotNet/RateLimits.cs:10]: Change all setters to be internal

File: OpenAI-DotNet/Completions/CompletionResult.cs (hunk 3 above)
Comment (StephenHodgson) [reply OpenAI-DotNet/Completions/CompletionResult.cs:52]: Move `RateLimits` to `BaseResponse`

File: OpenAI-DotNet/IUseRateLimits.cs [hunk 8]
```diff
@@ -0,0 +1,11 @@
+﻿using System.Text.Json.Serialization;
+
+namespace OpenAI
+{
+
+    public interface IUseRateLimits
```
Comment (StephenHodgson) [inline OpenAI-DotNet/IUseRateLimits.cs:6]: Remove this interface all together and just use `BaseResponse`

File: OpenAI-DotNet/Chat/ChatResponse.cs (hunk 4 above)
Comment (StephenHodgson) [reply OpenAI-DotNet/Chat/ChatResponse.cs:73]: Move to `BaseResponse`

File: OpenAI-DotNet/Extensions/ResponseExtensions.cs [hunk 9]
```diff
@@ -15,6 +15,14 @@ internal static class ResponseExtensions
         private const string RequestId = "X-Request-ID";
         private const string Organization = "Openai-Organization";
         private const string ProcessingTime = "Openai-Processing-Ms";
+        private const string OpenAIVersion = "openai-version";
+
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Extensions/ResponseExtensions.cs:19]: Remove whitespace

File: OpenAI-DotNet/Chat/ChatEndpoint.cs [hunk 10]
```diff
@@ -33,7 +33,9 @@ public async Task<ChatResponse> GetCompletionAsync(ChatRequest chatRequest, Canc
             var jsonContent = JsonSerializer.Serialize(chatRequest, OpenAIClient.JsonSerializationOptions).ToJsonStringContent(EnableDebug);
             var response = await Api.Client.PostAsync(GetUrl("/completions"), jsonContent, cancellationToken).ConfigureAwait(false);
             var responseAsString = await response.ReadAsStringAsync(EnableDebug, cancellationToken).ConfigureAwait(false);
-            return response.DeserializeResponse<ChatResponse>(responseAsString, OpenAIClient.JsonSerializationOptions);
+            var chatResponse = response.DeserializeResponse<ChatResponse>(responseAsString, OpenAIClient.JsonSerializationOptions);
+            chatResponse.SetResponseData(response.Headers);
+            return chatResponse;
         }
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Chat/ChatEndpoint.cs:39]: Revert and make sure we're calling `Response.Extensions.DeserializeResponse<T>`

**Review by StephenHodgson — CHANGES_REQUESTED** [review]
See comments

File: OpenAI-DotNet/BaseResponse.cs [hunk 11]
```diff
@@ -22,5 +22,12 @@ public abstract class BaseResponse
         /// </summary>
         [JsonIgnore]
         public string RequestId { get; internal set; }
+
+        /// <summary>
+        /// The version of the API used to generate this response, as reported in the response headers.
+        /// </summary>
+        [JsonIgnore]
+        public string OpenAIVersion { get; set; }
```
Comment (StephenHodgson) [inline OpenAI-DotNet/BaseResponse.cs:30]: Internal set only

**Review by StephenHodgson — COMMENTED** [review]
.

File: OpenAI-DotNet/RateLimits.cs [hunk 12]
```diff
@@ -0,0 +1,37 @@
+﻿namespace OpenAI
+{
+
+    public class RateLimits
```
Comment (StephenHodgson) [inline OpenAI-DotNet/RateLimits.cs:4]: Remove this completely and move all properties into `BaseResponse`

File: OpenAI-DotNet-Tests/TestFixture_03_Chat.cs [hunk 13]
```diff
@@ -678,5 +679,30 @@ public async Task Test_11_GetChatVisionStreaming()
             Console.WriteLine($"{result.FirstChoice.Message.Role}: {result.FirstChoice.Message.Content} | Finish Reason: {result.FirstChoice.FinishDetails}");
             result.GetUsage();
         }
+
+        [Test]
+        public async Task Test_12_ReadRateLimitHeaders()
```
Comment (StephenHodgson) [inline OpenAI-DotNet-Tests/TestFixture_03_Chat.cs:684]: I don't think we need a new test for this. Just add logging to the one of the existing tests

File: OpenAI-DotNet/BaseResponse.cs [hunk 14]
```diff
@@ -22,5 +22,49 @@ public abstract class BaseResponse
...
+        /// The remaining number of requests that are permitted before exhausting the rate limit.
+        /// </summary>
+        [JsonIgnore]
+        public int RemainingRequests { get; internal set; }
+
+        /// <summary>
+        /// The remaining number of tokens that are permitted before exhausting the rate limit.
+        /// </summary>
+        [JsonIgnore]
+        public int RemainingTokens { get; internal set; }
```
Comment (StephenHodgson) [inline OpenAI-DotNet/BaseResponse.cs:55]: lets make all these nullable 
```suggestion
        public int? RemainingTokens { get; internal set; }
```

File: OpenAI-DotNet/BaseResponse.cs [hunk 15]
```diff
@@ -22,5 +22,49 @@ public abstract class BaseResponse
...
+        /// The maximum number of tokens that are permitted before exhausting the rate limit.
+        /// </summary>
+        [JsonIgnore]
+        public int LimitTokens { get; internal set; }
+
+        /// <summary>
+        /// The remaining number of requests that are permitted before exhausting the rate limit.
+        /// </summary>
+        [JsonIgnore]
+        public int RemainingRequests { get; internal set; }
```
Comment (StephenHodgson) [inline OpenAI-DotNet/BaseResponse.cs:49]:
```suggestion
        public int? RemainingRequests { get; internal set; }
```

File: OpenAI-DotNet/BaseResponse.cs [hunk 16]
```diff
@@ -22,5 +22,49 @@ public abstract class BaseResponse
...
+        /// </summary>
+
+        [JsonIgnore] 
+        public int LimitRequests { get; internal set; }
+
+        /// <summary>
+        /// The maximum number of tokens that are permitted before exhausting the rate limit.
+        /// </summary>
+        [JsonIgnore]
+        public int LimitTokens { get; internal set; }
```
Comment (StephenHodgson) [inline OpenAI-DotNet/BaseResponse.cs:43]:
```suggestion
        public int? LimitTokens { get; internal set; }
```

File: OpenAI-DotNet/BaseResponse.cs [hunk 17]
```diff
@@ -22,5 +22,49 @@ public abstract class BaseResponse
...
+        /// </summary>
+        [JsonIgnore]
+        public string OpenAIVersion { get; internal set; }
+
+        /// <summary>
+        /// The maximum number of requests that are permitted before exhausting the rate limit.
+        /// </summary>
+
+        [JsonIgnore] 
+        public int LimitRequests { get; internal set; }
```
Comment (StephenHodgson) [inline OpenAI-DotNet/BaseResponse.cs:37]:
```suggestion
        public int? LimitRequests { get; internal set; }
```

File: OpenAI-DotNet/Extensions/ResponseExtensions.cs [hunk 18]
```diff
@@ -37,10 +44,51 @@ internal static void SetResponseData(this BaseResponse response, HttpResponseHea
             }
 
             if (headers.TryGetValues(ProcessingTime, out var processingTimeString) &&
-                double.TryParse(processingTimeString.First(), NumberStyles.AllowDecimalPoint, numberFormatInfo, out var processingTime))
+                double.TryParse(processingTimeString.First(), NumberStyles.AllowDecimalPoint, numberFormatInfo,
+                    out var processingTime))
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Extensions/ResponseExtensions.cs]: revert
```suggestion
                double.TryParse(processingTimeString.First(), NumberStyles.AllowDecimalPoint, numberFormatInfo, out var processingTime))
```

**Review by chsword — COMMENTED** [review]
I have formatted and modified the code as requested.

**Comment by StephenHodgson** [comment]
Thanks! we're close now :) 

I think we can make the int values nullable and we're good to go 😎

**Comment by chsword** [comment]
> Thanks! we're close now :)
> 
> I think we can make the int values nullable and we're good to go 😎

I just submitted the code.

File: OpenAI-DotNet/OpenAI-DotNet.csproj [hunk 19]
```diff
@@ -17,8 +17,11 @@ More context [on Roger Pincombe's blog](https://rogerpincombe.com/openai-dotnet-
     <RepositoryUrl>https://github.com/RageAgainstThePixel/OpenAI-DotNet</RepositoryUrl>
     <PackageTags>OpenAI, AI, ML, API, gpt-4, gpt-3.5-tubo, gpt-3, chatGPT, chat-gpt, gpt-2, gpt, dall-e-2, dall-e-3</PackageTags>
     <Title>OpenAI API</Title>
-    <Version>7.2.2</Version>
-    <PackageReleaseNotes>Version 7.2.2
+    <Version>7.2.3</Version>
+    <PackageReleaseNotes>Version 7.2.3
+- Added support for reading RateLimit information from the Headers in Completions and Chat endpoints.
+- Updated chatEndpoint.GetCompletionAsync to handle Headers processing.
```
Comment (StephenHodgson) [inline OpenAI-DotNet/OpenAI-DotNet.csproj:23]:
```suggestion
- Added support for reading RateLimit information from the Headers
```

File: OpenAI-DotNet-Tests/TestFixture_03_Chat.cs [hunk 20]
```diff
@@ -35,6 +35,14 @@ public async Task Test_01_GetChatCompletion()
...
 
             result.GetUsage();
+
+            Console.WriteLine(result.LimitRequests);
+            Console.WriteLine(result.RemainingRequests);
+            Console.WriteLine(result.ResetRequests);
+            Console.WriteLine(result.LimitTokens);
+            Console.WriteLine(result.RemainingTokens);
+            Console.WriteLine(result.ResetTokens);
+
```
Comment (StephenHodgson) [inline OpenAI-DotNet-Tests/TestFixture_03_Chat.cs:45]:
```suggestion
```

File: OpenAI-DotNet/BaseResponse.cs [hunk 21]
```diff
@@ -22,5 +22,49 @@ public abstract class BaseResponse
...
+        /// </summary>
+        [JsonIgnore]
+        public string ResetRequests { get; internal set; }
+
+        /// <summary>
+        /// The time until the rate limit (based on tokens) resets to its initial state.
+        /// </summary>
+        [JsonIgnore]
+        public string ResetTokens { get; internal set; }
+
```
Comment (StephenHodgson) [inline OpenAI-DotNet/BaseResponse.cs:68]:
```suggestion
```

**Comment by StephenHodgson** [comment]
one last request and we will merge. 

Please update readme with changes from https://github.com/RageAgainstThePixel/OpenAI-DotNet/pull/163/files

**Review by StephenHodgson — APPROVED** [review]

**Pull request merged**